compiler/rustc_trait_selection/src/traits/normalize.rs

```diff
@@ -75,7 +75,11 @@ impl<'tcx> At<'_, 'tcx> {
             let value = self
                 .normalize(value)
                 .into_value_registering_obligations(self.infcx, &mut *fulfill_cx);
-            let errors = fulfill_cx.select_all_or_error(self.infcx);
+            let mut errors = fulfill_cx.select_all_or_error(self.infcx);
+            // Drain pending obligations too, since deep normalization may happen
+            // in a loop and we don't want to trigger the assertion on the next
+            // iteration due to pending obligations we've left over.
+            errors.extend(fulfill_cx.collect_remaining_errors(self.infcx));
```

Comment:
hmm, `fn select_all_or_error` is currently implemented as
```rust
    fn select_all_or_error(&mut self, infcx: &InferCtxt<'tcx>) -> Vec<E> {
        let errors = self.select_where_possible(infcx);
        if !errors.is_empty() {
            return errors;
        }

        self.collect_remaining_errors(infcx)
    }
```
it feels a bit footgunny to not add the overflowing obligations if there are other errors. I think changing the `if errors.is_empty()` branch to also add the remaining errors would be better :thinking:

Reply:
or no, in case `select_where_possible` returns an error, we *want* to ignore overflow errors. I think we also want to ignore these ambiguity errors here, so maybe change this to
```rust
            let errors = fulfill_cx.select_all_or_error(self.infcx);
            if errors.is_empty() {
                Ok(self.infcx.resolve_vars_if_possible(value))
            } else {
                // In case normalization resulted in an error, we want to
                // drain any still ambiguous goal to avoid triggering the
                // assertion on the next iteration.
                let _ = fulfill_cx.collect_remaining_errors(self.infcx));
                Err(errors)
            }
```

Reply:
alternatively we could drop the remaining issues in the `!errors.is_empty()`  branch in `select_all_or_error` itself? :thinking:

Reply:
> alternatively we could drop the remaining obligations in the `!errors.is_empty()` branch in `select_all_or_error` itself? 🤔

Hm, feels like a much larger and more subtle change. I'd rather keep it located to deep normalization.

Reply:
> or no, in case select_where_possible returns an error, we want to ignore overflow and ambiguity errors. I think we also want to ignore these ambiguity errors here, so maybe change this to [...]

Functionally equivalent to what I have right now, but I guess alright.

Reply:
> Functionally equivalent to what I have right now, but I guess alright.

Correction -- not functionally equivalent, but doesn't matter for UI tests. I originally had `let _ = fulfill_cx.collect_remaining_errors(self.infcx))` on the good path of deep norm too, but it felt spooky. I guess it's fine though.